File: packages/brain-py/maia_brain/provenance.py
"""Read-only provenance graph helpers for ACP event streams."""

from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any

from maia_acp.types import (
    ACPArtifact,
    ACPEvent,
    ACPMessage,
    ACPProvenanceGraph,
    ACPReview,
    ProvenanceClaim,
    ProvenanceContradiction,
    ProvenanceSourceRef,
)

URL_PATTERN = re.compile(r"\bhttps?://[^\s)]+", re.IGNORECASE)
SENTENCE_SPLIT = re.compile(r"(?<=[.!?])\s+|\n+")
NUMBER_PATTERN = re.compile(r"-?\d+(?:\.\d+)?(?:%|[mbk]| billion| million| thousand)?", re.IGNORECASE)


def _compact_text(text: str) -> str:
    return re.sub(r"\s+", " ", str(text or "")).strip()


def _slug(value: str) -> str:
    compact = _compact_text(value).lower()
    compact = re.sub(r"[^a-z0-9]+", "_", compact).strip("_")
    return compact[:60] or "item"


def _source_id(kind: str, seed: str) -> str:
    return f"{kind}_{_slug(seed)}"


def _tier_for_sources(source_count: int) -> str:
    if source_count >= 2:
        return "verified"
    if source_count == 1:
        return "supported"
    return "inferred"


def _confidence_for_tier(tier: str, source_count: int) -> float:
    if tier == "verified":
        return min(0.95, 0.72 + source_count * 0.08)
    if tier == "supported":
        return 0.68
    if tier == "inferred":
        return 0.48
    return 0.2


def _extract_urls(text: str) -> list[str]:
    return list(dict.fromkeys(URL_PATTERN.findall(str(text or ""))))


def _sentence_candidates(text: str, max_claims: int) -> list[str]:
    return [
        candidate
        for candidate in (_compact_text(part) for part in SENTENCE_SPLIT.split(_compact_text(text)))
        if 25 <= len(candidate) <= 280
    ][:max_claims]


def _claim_id(agent_id: str, event_id: str, text: str, index: int) -> str:
    return f"claim_{_slug(agent_id)}_{_slug(event_id)}_{index}_{_slug(text)[:18]}"


def _contradiction_id(claim_a_id: str, claim_b_id: str) -> str:
    ordered = sorted([claim_a_id, claim_b_id])
    return f"contradiction_{ordered[0]}_{ordered[1]}"


def _synthetic_event_id(event: ACPEvent, index: int) -> str:
    return f"{event.run_id}_{event.sequence or index}_{event.event_type}"


def _unique_source_refs(source_refs: list[ProvenanceSourceRef]) -> list[ProvenanceSourceRef]:
    seen: set[str] = set()
    result: list[ProvenanceSourceRef] = []
    for source_ref in source_refs:
        key = f"{source_ref.kind}:{source_ref.source_id}:{source_ref.uri or ''}:{source_ref.artifact_id or ''}:{source_ref.event_id or ''}"
        if key in seen:
            continue
        seen.add(key)
        result.append(source_ref)
    return result


def _message_sources(payload: ACPMessage, event: ACPEvent, event_id: str) -> list[ProvenanceSourceRef]:
    refs = [
        ProvenanceSourceRef(
            source_id=_source_id("url", url),
            kind="url",
            uri=url,
            title=url,
            event_id=event_id,
            accessed_at=event.timestamp,
        )
        for url in _extract_urls(payload.content)
    ]
    refs.extend(
        ProvenanceSourceRef(
            source_id=_source_id("artifact", artifact.artifact_id),
            kind="artifact",
            artifact_id=artifact.artifact_id,
            title=artifact.title,
            uri=artifact.content_url,
            event_id=event_id,
            excerpt=_compact_text(artifact.content)[:160],
            accessed_at=event.timestamp,
        )
        for artifact in (payload.artifacts or [])
    )
    return _unique_source_refs(refs)


def _artifact_sources(payload: ACPArtifact, event: ACPEvent, event_id: str) -> list[ProvenanceSourceRef]:
    refs = [
        ProvenanceSourceRef(
            source_id=_source_id("document" if payload.kind == "pdf" else "artifact", payload.artifact_id),
            kind="document" if payload.kind == "pdf" else "artifact",
            artifact_id=payload.artifact_id,
            title=payload.title,
            uri=payload.content_url,
            event_id=event_id,
            excerpt=_compact_text(payload.content)[:160],
            accessed_at=event.timestamp,
        )
    ]
    refs.extend(
        ProvenanceSourceRef(
            source_id=_source_id("url", url),
            kind="url",
            uri=url,
            title=url,
            event_id=event_id,
            accessed_at=event.timestamp,
        )
        for url in _extract_urls(payload.content)
    )
    return _unique_source_refs(refs)


def _review_sources(payload: ACPReview, event_id: str, timestamp: str) -> list[ProvenanceSourceRef]:
    return [
        ProvenanceSourceRef(
            source_id=_source_id("reasoning", f"{payload.reviewer}_{payload.author}_{payload.artifact_id or payload.feedback or 'review'}"),
            kind="reasoning",
            title=f"Review by {payload.reviewer}",
            event_id=event_id,
            excerpt=_compact_text(payload.feedback or payload.revision_instructions or payload.verdict),
            accessed_at=timestamp,
        )
    ]


def extract_claims_from_text(
    *,
    text: str,
    agent_id: str,
    event_id: str,
    source_refs: list[ProvenanceSourceRef] | None = None,
    metadata: dict[str, Any] | None = None,
    max_claims: int = 4,
) -> list[ProvenanceClaim]:
    refs = _unique_source_refs(source_refs or [])
    tier = _tier_for_sources(len(refs))
    return [
        ProvenanceClaim(
            claim_id=_claim_id(agent_id, event_id, sentence, index),
            text=sentence,
            agent_id=agent_id,
            tier=tier,
            confidence=_confidence_for_tier(tier, len(refs)),
            source_refs=refs,
            supports=[],
            contradicts=[],
            metadata=metadata or {},
        )
        for index, sentence in enumerate(_sentence_candidates(text, max_claims))
    ]


def _normalized_pattern(text: str) -> str:
    normalized = _compact_text(text).lower()
    normalized = NUMBER_PATTERN.sub("#", normalized)
    normalized = re.sub(r"[^a-z# ]+", " ", normalized)
    return re.sub(r"\s+", " ", normalized).strip()


def _numbers(text: str) -> list[str]:
    return [value.lower() for value in NUMBER_PATTERN.findall(_compact_text(text))]


def detect_contradictions(claims: list[ProvenanceClaim]) -> list[ProvenanceContradiction]:
    contradictions: list[ProvenanceContradiction] = []
    for index, claim_a in enumerate(claims):
        for claim_b in claims[index + 1 :]:
            if _normalized_pattern(claim_a.text) != _normalized_pattern(claim_b.text):
                continue
            numbers_a = _numbers(claim_a.text)
            numbers_b = _numbers(claim_b.text)
            if not numbers_a or not numbers_b or numbers_a == numbers_b:
                continue
            contradictions.append(
                ProvenanceContradiction(
                    contradiction_id=_contradiction_id(claim_a.claim_id, claim_b.claim_id),
                    claim_a_id=claim_a.claim_id,
                    claim_b_id=claim_b.claim_id,
                    status="unresolved",
                )
            )
    return contradictions


def build_provenance_graph(events: list[ACPEvent]) -> ACPProvenanceGraph:
    claims: list[ProvenanceClaim] = []
    explicit_graphs: list[ACPProvenanceGraph] = []

    for index, event in enumerate(events):
        event_id = _synthetic_event_id(event, index)
        if event.event_type == "provenance":
            explicit_graphs.append(event.as_provenance())
            continue
        if event.event_type == "message":
            payload = event.as_message()
            claims.extend(
                extract_claims_from_text(
                    text=payload.content,
                    agent_id=payload.from_agent,
                    event_id=event_id,
                    source_refs=_message_sources(payload, event, event_id),
                    metadata={
                        "event_type": event.event_type,
                        "message_intent": payload.intent,
                        "thread_id": (payload.context or {}).get("thread_id"),
                    },
                )
            )
            continue
        if event.event_type == "artifact":
            payload = event.as_artifact()
            claims.extend(
                extract_claims_from_text(
                    text=payload.content,
                    agent_id=event.agent_id,
                    event_id=event_id,
                    source_refs=_artifact_sources(payload, event, event_id),
                    metadata={
                        "event_type": event.event_type,
                        "artifact_kind": payload.kind,
                        "artifact_id": payload.artifact_id,
                    },
                    max_claims=5,
                )
            )
            continue
        if event.event_type == "review":
            payload = event.as_review()
            review_text = _compact_text(". ".join(part for part in [payload.feedback, payload.revision_instructions] if part))
            if review_text:
                claims.extend(
                    extract_claims_from_text(
                        text=review_text,
                        agent_id=payload.reviewer,
                        event_id=event_id,
                        source_refs=_review_sources(payload, event_id, event.timestamp),
                        metadata={
                            "event_type": event.event_type,
                            "verdict": payload.verdict,
                            "author": payload.author,
                        },
                        max_claims=3,
                    )
                )

    merged_claims = [claim for graph in explicit_graphs for claim in graph.claims] + claims
    contradictions = [item for graph in explicit_graphs for item in graph.contradictions] + detect_contradictions(merged_claims)

    claim_map: dict[str, ProvenanceClaim] = {
        claim.claim_id: claim.model_copy(deep=True)
        for claim in merged_claims
    }
    for contradiction in contradictions:
        if contradiction.claim_a_id in claim_map:
            claim_map[contradiction.claim_a_id].contradicts.append(contradiction.claim_b_id)
        if contradiction.claim_b_id in claim_map:
            claim_map[contradiction.claim_b_id].contradicts.append(contradiction.claim_a_id)

    deduped_contradictions = list({item.contradiction_id: item for item in contradictions}.values())
    final_claims = []
    for claim in claim_map.values():
        clone = claim.model_copy(deep=True)
        clone.contradicts = list(dict.fromkeys(clone.contradicts))
        final_claims.append(clone)

    return ACPProvenanceGraph(
        graph_id=explicit_graphs[-1].graph_id if explicit_graphs else f"graph_{events[-1].run_id if events else 'unknown'}",
        run_id=explicit_graphs[-1].run_id if explicit_graphs else (events[-1].run_id if events else ""),
        claims=final_claims,
        contradictions=deduped_contradictions,
    )


def stale_claims(graph: ACPProvenanceGraph, max_age_days: int = 90) -> list[ProvenanceClaim]:
    max_age_seconds = max_age_days * 24 * 60 * 60
    now = datetime.now(timezone.utc)
    stale: list[ProvenanceClaim] = []
    for claim in graph.claims:
        for source_ref in claim.source_refs:
            stamp = source_ref.published_at or source_ref.accessed_at
            if not stamp:
                continue
            try:
                parsed = datetime.fromisoformat(stamp.replace("Z", "+00:00"))
            except ValueError:
                continue
            if (now - parsed).total_seconds() > max_age_seconds:
                stale.append(claim)
                break
    return stale

File: packages/brain-runtime/src/provenance.ts
import type {
  ACPArtifact,
  ACPEvent,
  ACPMessage,
  ACPProvenanceGraph,
  ACPReview,
  ProvenanceClaim,
  ProvenanceContradiction,
  ProvenanceSourceRef,
  ProvenanceTier,
} from "@maia/acp";

export interface ClaimExtractionResult {
  claims: ProvenanceClaim[];
}

const URL_PATTERN = /\bhttps?:\/\/[^\s)]+/gi;
const SENTENCE_SPLIT = /(?<=[.!?])\s+|\n+/g;
const NUMBER_PATTERN = /-?\d+(?:\.\d+)?(?:%|[mbk]| billion| million| thousand)?/gi;

function compactText(text: string): string {
  return String(text || "").replace(/\s+/g, " ").trim();
}

function slug(value: string): string {
  return compactText(value)
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "")
    .slice(0, 60) || "item";
}

function sourceId(kind: ProvenanceSourceRef["kind"], seed: string): string {
  return `${kind}_${slug(seed)}`;
}

function uniqueSourceRefs(sourceRefs: ProvenanceSourceRef[]): ProvenanceSourceRef[] {
  const seen = new Set<string>();
  const result: ProvenanceSourceRef[] = [];
  for (const sourceRef of sourceRefs) {
    const key = `${sourceRef.kind}:${sourceRef.source_id}:${sourceRef.uri ?? ""}:${sourceRef.artifact_id ?? ""}:${sourceRef.event_id ?? ""}`;
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    result.push(sourceRef);
  }
  return result;
}

function tierForSources(sourceCount: number): ProvenanceTier {
  if (sourceCount >= 2) {
    return "verified";
  }
  if (sourceCount === 1) {
    return "supported";
  }
  return "inferred";
}

function confidenceForTier(tier: ProvenanceTier, sourceCount: number): number {
  if (tier === "verified") {
    return Math.min(0.95, 0.72 + sourceCount * 0.08);
  }
  if (tier === "supported") {
    return 0.68;
  }
  if (tier === "inferred") {
    return 0.48;
  }
  return 0.2;
}

function extractUrls(text: string): string[] {
  return Array.from(new Set(String(text || "").match(URL_PATTERN) ?? []));
}

function sentenceCandidates(text: string, maxClaims: number): string[] {
  const parts = compactText(text)
    .split(SENTENCE_SPLIT)
    .map((part) => compactText(part))
    .filter((part) => part.length >= 25 && part.length <= 280);
  return parts.slice(0, maxClaims);
}

function claimId(agentId: string, eventId: string, text: string, index: number): string {
  return `claim_${slug(agentId)}_${slug(eventId)}_${index}_${slug(text).slice(0, 18)}`;
}

function contradictionId(claimAId: string, claimBId: string): string {
  const ordered = [claimAId, claimBId].sort();
  return `contradiction_${ordered.join("_")}`;
}

function syntheticEventId(event: ACPEvent, index: number): string {
  return `${event.run_id}_${event.sequence ?? index}_${event.event_type}`;
}

function buildMessageSources(payload: ACPMessage, event: ACPEvent, eventId: string): ProvenanceSourceRef[] {
  const urls = extractUrls(payload.content).map((url) => ({
    source_id: sourceId("url", url),
    kind: "url" as const,
    uri: url,
    title: url,
    event_id: eventId,
    accessed_at: event.timestamp,
  }));
  const artifacts = (payload.artifacts ?? []).map((artifact) => ({
    source_id: sourceId("artifact", artifact.artifact_id),
    kind: "artifact" as const,
    artifact_id: artifact.artifact_id,
    title: artifact.title,
    uri: artifact.content_url,
    event_id: eventId,
    excerpt: compactText(artifact.content).slice(0, 160),
    accessed_at: event.timestamp,
  }));
  return uniqueSourceRefs([...urls, ...artifacts]);
}

function buildArtifactSources(payload: ACPArtifact, event: ACPEvent, eventId: string): ProvenanceSourceRef[] {
  const refs: ProvenanceSourceRef[] = [
    {
      source_id: sourceId(payload.kind === "pdf" ? "document" : "artifact", payload.artifact_id),
      kind: payload.kind === "pdf" ? "document" : "artifact",
      artifact_id: payload.artifact_id,
      title: payload.title,
      uri: payload.content_url,
      event_id: eventId,
      excerpt: compactText(payload.content).slice(0, 160),
      accessed_at: event.timestamp,
    },
  ];
  for (const url of extractUrls(payload.content)) {
    refs.push({
      source_id: sourceId("url", url),
      kind: "url",
      uri: url,
      title: url,
      event_id: eventId,
      accessed_at: event.timestamp,
    });
  }
  return uniqueSourceRefs(refs);
}

function buildReviewSources(payload: ACPReview, eventId: string, timestamp: string): ProvenanceSourceRef[] {
  return [
    {
      source_id: sourceId("reasoning", `${payload.reviewer}_${payload.author}_${payload.artifact_id ?? payload.feedback ?? "review"}`),
      kind: "reasoning",
      title: `Review by ${payload.reviewer}`,
      event_id: eventId,
      excerpt: compactText(payload.feedback ?? payload.revision_instructions ?? payload.verdict),
      accessed_at: timestamp,
    },
  ];
}

function normalizeClaimPattern(text: string): string {
  const withoutUrls = compactText(text).replace(URL_PATTERN, " ");
  return withoutUrls
    .toLowerCase()
    .replace(NUMBER_PATTERN, "#")
    .replace(/[^a-z# ]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function extractNumbers(text: string): string[] {
  return (compactText(text).match(NUMBER_PATTERN) ?? []).map((value) => value.toLowerCase());
}

export function extractClaimsFromText(input: {
  text: string;
  agentId: string;
  eventId: string;
  sourceRefs?: ProvenanceSourceRef[];
  metadata?: Record<string, unknown>;
  maxClaims?: number;
}): ClaimExtractionResult {
  const claims = sentenceCandidates(input.text, input.maxClaims ?? 4).map((sentence, index) => {
    const sourceRefs = uniqueSourceRefs(input.sourceRefs ?? []);
    const tier = tierForSources(sourceRefs.length);
    return {
      claim_id: claimId(input.agentId, input.eventId, sentence, index),
      text: sentence,
      agent_id: input.agentId,
      tier,
      confidence: confidenceForTier(tier, sourceRefs.length),
      source_refs: sourceRefs,
      supports: [],
      contradicts: [],
      metadata: input.metadata ?? {},
    } satisfies ProvenanceClaim;
  });
  return { claims };
}

export function detectContradictions(claims: ProvenanceClaim[]): ProvenanceContradiction[] {
  const contradictions: ProvenanceContradiction[] = [];
  for (let index = 0; index < claims.length; index++) {
    for (let inner = index + 1; inner < claims.length; inner++) {
      const claimA = claims[index];
      const claimB = claims[inner];
      if (claimA.claim_id === claimB.claim_id) {
        continue;
      }
      const patternA = normalizeClaimPattern(claimA.text);
      const patternB = normalizeClaimPattern(claimB.text);
      if (!patternA || patternA !== patternB) {
        continue;
      }
      const numbersA = extractNumbers(claimA.text);
      const numbersB = extractNumbers(claimB.text);
      if (numbersA.length === 0 || numbersB.length === 0) {
        continue;
      }
      if (numbersA.join("|") === numbersB.join("|")) {
        continue;
      }
      contradictions.push({
        contradiction_id: contradictionId(claimA.claim_id, claimB.claim_id),
        claim_a_id: claimA.claim_id,
        claim_b_id: claimB.claim_id,
        status: "unresolved",
      });
    }
  }
  return contradictions;
}

export function buildProvenanceGraph(events: ACPEvent[]): ACPProvenanceGraph {
  const claims: ProvenanceClaim[] = [];
  const explicitGraphs: ACPProvenanceGraph[] = [];

  events.forEach((event, index) => {
    const eventId = syntheticEventId(event, index);
    if (event.event_type === "provenance") {
      explicitGraphs.push(event.payload as ACPProvenanceGraph);
      return;
    }

    if (event.event_type === "message") {
      const payload = event.payload as ACPMessage;
      claims.push(
        ...extractClaimsFromText({
          text: payload.content,
          agentId: payload.from,
          eventId,
          sourceRefs: buildMessageSources(payload, event, eventId),
          metadata: {
            event_type: event.event_type,
            message_intent: payload.intent,
            thread_id: payload.context?.thread_id,
          },
        }).claims,
      );
      return;
    }

    if (event.event_type === "artifact") {
      const payload = event.payload as ACPArtifact;
      claims.push(
        ...extractClaimsFromText({
          text: payload.content,
          agentId: event.agent_id,
          eventId,
          sourceRefs: buildArtifactSources(payload, event, eventId),
          metadata: {
            event_type: event.event_type,
            artifact_kind: payload.kind,
            artifact_id: payload.artifact_id,
          },
          maxClaims: 5,
        }).claims,
      );
      return;
    }

    if (event.event_type === "review") {
      const payload = event.payload as ACPReview;
      const reviewText = compactText([payload.feedback, payload.revision_instructions].filter(Boolean).join(". "));
      if (!reviewText) {
        return;
      }
      claims.push(
        ...extractClaimsFromText({
          text: reviewText,
          agentId: payload.reviewer,
          eventId,
          sourceRefs: buildReviewSources(payload, eventId, event.timestamp),
          metadata: {
            event_type: event.event_type,
            verdict: payload.verdict,
            author: payload.author,
          },
          maxClaims: 3,
        }).claims,
      );
    }
  });

  const explicitClaims = explicitGraphs.flatMap((graph) => graph.claims ?? []);
  const explicitContradictions = explicitGraphs.flatMap((graph) => graph.contradictions ?? []);
  const mergedClaims = [...explicitClaims, ...claims];
  const contradictions = [
    ...explicitContradictions,
    ...detectContradictions(mergedClaims),
  ];

  const contradictionMap = new Map<string, ProvenanceClaim>();
  for (const claim of mergedClaims) {
    contradictionMap.set(claim.claim_id, { ...claim, contradicts: [...claim.contradicts] });
  }
  for (const contradiction of contradictions) {
    contradictionMap.get(contradiction.claim_a_id)?.contradicts.push(contradiction.claim_b_id);
    contradictionMap.get(contradiction.claim_b_id)?.contradicts.push(contradiction.claim_a_id);
  }

  return {
    graph_id: explicitGraphs.at(-1)?.graph_id ?? `graph_${events.at(-1)?.run_id ?? "unknown"}`,
    run_id: explicitGraphs.at(-1)?.run_id ?? events.at(-1)?.run_id ?? "",
    claims: Array.from(contradictionMap.values()).map((claim: ProvenanceClaim) => ({
      ...claim,
      contradicts: Array.from(new Set(claim.contradicts)),
    })),
    contradictions: Array.from(
      new Map<string, ProvenanceContradiction>(
        contradictions.map((contradiction: ProvenanceContradiction) => [contradiction.contradiction_id, contradiction]),
      ).values(),
    ),
  };
}

export function staleClaims(graph: ACPProvenanceGraph, maxAgeDays = 90): ProvenanceClaim[] {
  const maxAgeMs = maxAgeDays * 24 * 60 * 60 * 1000;
  const now = Date.now();
  return graph.claims.filter((claim: ProvenanceClaim) =>
    claim.source_refs.some((sourceRef: ProvenanceSourceRef) => {
      const stamp = sourceRef.published_at ?? sourceRef.accessed_at;
      if (!stamp) {
        return false;
      }
      const time = Date.parse(stamp);
      if (Number.isNaN(time)) {
        return false;
      }
      return now - time > maxAgeMs;
    }),
  );
}
